### PR TITLE
[FEATURE] Add a simple string cache

### PR DESCRIPTION
**Pull request by oliverklee**
This class then will later be used to replace the current caching
that is based on nested arrays.

Part of #144.

### CODE REVIEW

**Review by JakeQZ — COMMENTED** [review]
Could the class implement [`ArrayAccess`](https://www.php.net/manual/en/class.arrayaccess.php) (instead of the `set`/`get`/`has` methods)?  This would also mean fewer changes to the existing code which could continue to get, set and test the cached values using the `[]` operator and `isset`.

**Comment by oliverklee** [comment]
As far as I understand it, this would mostly remove the benefit of this class over using a plain array:
- typing the keys to strings
- typing the values to strings
- having more "speaking" method names.

So I'd prefer to now implement `ArrayAccess` for these reasons.

**Comment by JakeQZ** [comment]
> 
> 
>     * typing the keys to strings
> 
>     * typing the values to strings

Yeah, I wondered about that, and can't think of or find a solution.  Unlike C++, PHP does not have templates so it's not possible to `implment ArrayAccess<string, string>`.

File: src/Caching/SimpleStringCache.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * ```php
+ * $cache = new SimpleStringCache();
+ * $cache->set($key, $value);
+ * …
+ * if ($cache->has($key) {
+ *     $cachedValue = $cache->get($value);
+ * }
+ * ```
+ */
```
Comment (JakeQZ) [inline src/Caching/SimpleStringCache.php:20]: Should the class be `@internal`?

File: src/Caching/SimpleStringCache.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    private $values = [];
+
+    /**
+     * Checks whether there is an entry stored for the given key.
+     *
+     * @param string $key the key to check; must not be empty
+     *
+     * @throws \InvalidArgumentException
+     */
```
Comment (JakeQZ) [inline src/Caching/SimpleStringCache.php:34]: Should there be `@return bool` or is this superfluous as it does not provide any additional information?

File: src/Caching/SimpleStringCache.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    public function set(string $key, string $value): void
+    {
+        $this->checkForNonEmptyKey($key);
+
+        $this->values[$key] = $value;
+    }
+
+    /**
+     * @throws \InvalidArgumentException
+     */
```
Comment (JakeQZ) [inline src/Caching/SimpleStringCache.php:78]: Again, I assume `@param string $key` would be superfluous so is not needed.

File: src/Caching/SimpleStringCache.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    private function checkForNonEmptyKey(string $key): void
```
Comment (JakeQZ) [inline src/Caching/SimpleStringCache.php:79]: Wondering about the method name.  Would something like `assertNotEmptyKey` be more descriptive?  If `assert` may sound like conflating with test methods, perhaps `verify` or `confirm` (or something like) - these verbs all imply that an empty key will be rejected, whereas `checkFor` doesn't really say what will be done when the key is/isn't found to be empty.

File: tests/Unit/Caching/SimpleStringCacheTest.php [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        $this->expectExceptionCode(1625995840);
+        $this->expectExceptionMessage('Please provide a non-empty key.');
+
+        $this->subject->has('');
+    }
+
+    /**
+     * @test
+     */
+    public function hasInitiallyIsFalseForAnyKey(): void
```
Comment (JakeQZ) [inline tests/Unit/Caching/SimpleStringCacheTest.php:40]: Perhaps `Returns` rather than `Is` in the method name - ?

File: tests/Unit/Caching/SimpleStringCacheTest.php [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+        $this->subject->set($key, $value);
+
+        self::assertSame($value, $this->subject->get($key));
+    }
+
+    /**
+     * @test
+     */
+    public function getForKeyThatHasBeenSetTwiceReturnsValueThatHasBeenSetLast(): void
```
Comment (JakeQZ) [inline tests/Unit/Caching/SimpleStringCacheTest.php:130]: Weirdly, it feels more natural to say `getForKeyThatHasBeenSetTwiceReturnsValueThatWasSetLast` even though the verb tenses don't seem to agree.  Though it could probably be named `getForKeyThatHasBeenSetTwiceReturnsValueSetLast` to more concisely convey the same meaning and dodge the grammar issue.

File: tests/Unit/Caching/SimpleStringCacheTest.php [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+     */
+    public function setAndGetCanAlsoReturnEmptyString(): void
+    {
+        $key = 'hello';
+        $value = '';
+
+        $this->subject->set($key, $value);
+
+        self::assertSame($value, $this->subject->get($key));
+    }
```
Comment (JakeQZ) [inline tests/Unit/Caching/SimpleStringCacheTest.php:152]: I think this should be combined with `getForKeyThatHasBeenSetReturnsSetValue`, with a data provider providing a non-empty string and an empty string.

**Review by JakeQZ — CHANGES_REQUESTED** [review]
Looks generally good, and I can't think of any tests that are missing.

A few minor issues and questions.

**Comment by oliverklee** [comment]
@JakeQZ I immensely appreciate your detailed reviews! :heart: On it.

File: phpcs.xml [hunk 8]
```diff
@@ -39,6 +39,8 @@
   <rule ref="Squiz.Commenting.DocCommentAlignment"/>
   <rule ref="Squiz.Commenting.EmptyCatchComment"/>
   <rule ref="Squiz.Commenting.FunctionComment">
+    <exclude-pattern>*/tests/*</exclude-pattern>
+
```
[the remark this answers is not in the document]
Comment (oliverklee) [reply phpcs.xml:43]: As a separation between two kinds of `exclude`s: by pattern, and by name. I slightly prefer having the blank line here, but would be fine with removing it as well.

WDYT?

File: src/Caching/SimpleStringCache.php (hunk 2 above)
Comment (oliverklee) [reply src/Caching/SimpleStringCache.php:34]: I think it's superfluous.

Comment (JakeQZ) [reply src/Caching/SimpleStringCache.php:36]: OK, I kind-of thought so too.  (Also, we have recently removed insistence on superfluous DocBlock annotations from the sniffs.  We could perhaps even insist on not having superfluous DocBlock annotations if there is a sniff for it.)

File: phpcs.xml (hunk 8 above)
[the remark this answers is not in the document]
Comment (JakeQZ) [reply phpcs.xml:43]: Happy either way.  Just wanted to check it was deliberate and what the reason was.

**Review by JakeQZ — APPROVED** [review]
Lovely.  (Sorry for the delay getting back to you, been away quite a but and also busy on other projects.)
